src/poprox_recommender/test_json.py

```diff
@@ -0,0 +1,153 @@
+    mrr = 0
+    for rank, pred_item in enumerate(y_pred, start = 1):
+        if pred_item in relevant_items:
+            mrr += (1/rank)
+    return mrr/len(relevant_items)
+
+
+
+def compute_dcg(y_pred, y_true, k=5):
+    '''
+    y_pred: 排好的item顺序
+    y_true:[item-label]
+    '''
+    item_label = {}
+    for pair in y_true:
+        item_label[pair.split('-')[0]] = int(pair.split('-')[1])
+
+    y_true = np.array([item_label[item] for item in y_pred[:k]])
+
+    gains = 2 ** y_true - 1
+    discounts = np.log2(np.arange(len(y_true)) + 2)
+
+    return np.sum(gains / discounts)
+
+
+def compute_ndcg(y_pred, y_true, k):
+    '''
+    y_pred: 排好的item顺序
+    y_true:[item-label]
+    '''
+    # y_true2 作为best的item排序
+    y_true_score = [item.split('-')[1] for item in y_true]
+    order = np.argsort(y_true_score)[::-1]
+    y_true_item = [y_true[index].split('-')[0] for index in order]
+
+    best = compute_dcg(y_true_item, y_true, k)
+    actual = compute_dcg(y_pred, y_true, k)
+    return actual / best
+
+def recsys_metric(recommendations, row_index, news_struuid_ID ):
+    # recommendations {account id (uuid): LIST[Article]}
+    # use the url of Article
+    impressions_truth = pd.read_table("/home/XLL1713/POPROX/engine0/NRMS_bert/data/mind/large/val/behaviors.tsv",
+                                    header='infer',
+                                    usecols=range(5),
+                                    names=[
+                                        'impression_id', 'user', 'time',
+                                        'clicked_news', 'impressions'
+                                    ]).iloc[row_index].impressions.split() # LIST[ID-label]
+
+    account_id = list(recommendations.keys())[0]
+    recommended_list = recommendations[account_id]
+    recommended_list = [news_struuid_ID[item.url] for item in recommended_list]
+
+    single_mrr = compute_mrr(recommended_list, impressions_truth)
+    single_ndcg5 = compute_ndcg(recommended_list, impressions_truth, 5)
+    single_ndcg10 = compute_ndcg(recommended_list, impressions_truth, 10)
+
+    return single_ndcg5, single_ndcg10, single_mrr
+
+if __name__ == '__main__':
+    '''
+    For offline evaluation, set theta in mmr_diversity = 1
+    '''
+    MODEL, DEVICE = load_model()
+    TOKEN_MAPPING = 'distilbert-base-uncased'  # can be modified
+
+    with open('/home/XLL1713/POPROX/poprox-recommender/tests/news_uuid_ID.json', 'r') as json_file:
+        news_struuid_ID = json.load(json_file)
+
+    # load the mind test json file
+    with open('/home/XLL1713/POPROX/poprox-recommender/tests/mind_test.json', 'r') as json_file:
+        mind_data = json.load(json_file)
+
+    ndcg5 = []
+    ndcg10 = []
+    mrr = []
+
+    for impression_idx in range(len(mind_data)): # one by one
+
+        request_body = mind_data[impression_idx]
+
+        todays_articles = [Article.parse_obj(item) for item in request_body["todays_articles"]]
```

Comment:
This will totally work and I'm happy to go forward with this code as is, but just FYI you can also construct these objects directly like `Article(title=mind_data["some_title_key"], url=mind_data["the_url_key"])` which would probably allow you to skip the step of converting the MIND data to JSON first